### DISCUSSION TITLE
Interaction via ssh

### DISCUSSION THREAD

**Discussion by m13slash9** [neurodesk/neurodesk.github.io-hugo#421] [General]
Is it possible to interact with Neurodesk via ssh? If yes, how can the connection be enabled?

It would be nice to have a possibility to use some tools in Neurodesk by providing commands over ssh and data transfer over scp.
This could work both with Neurodesk installed locally and remotely, but at the moment I'm more interested in connecting to a local installation.

**Comment by stebo85** [comment]
Yes, this is possible. You can install Neurocommand locally (or remotely) and then connect to this setup via ssh. You can follow these instructions: https://www.neurodesk.org/docs/getting-started/neurocommand/

On the local setup you wouldn't even need SSH - you can call the containers directly once setup with neurocommand.

Another way would be to setup only the CVMFS container section of Neurodesk: https://www.neurodesk.org/docs/getting-started/neurocontainers/cvmfs/ (this can also be done remotely).

Let us know if this works for you. If not can you describe your setup in more detail (e.g. which operating systems are involved) and we can give more detailed guidance.

**Reply by m13slash9** [reply]
Thank you for the direction. I think I can try those.

At the moment I was hoping for a simpler setup: I have the NeuroDesk-App installed on Windows with Docker (and WSL2-integrated) in the background, which is a very easy (basically, two-step, very intuitive) setup for Windows. I was hoping I can ssh into the host that is run, when I select "Open Local Neurodesk". 

Same with Linux machines: I was hoping to use one of the containerized version of Neurodesk to have an easy installation, and still be able to ssh into it to.

I understand It might be redundant to use Neurodesktop in this manner, but the ease of installation is tempting.

**Reply by stebo85** [reply]
That could be possible, too. Can you try if this works for you: https://www.neurodesk.org/docs/getting-started/neurocommand/visual-studio-code/

it sounds very close to what you want to achieve

**Reply by m13slash9** [reply]
Thank you, but that seems to employ some tools very particular to VS.

I'm wondering something more general like ssh can be used. 
I guess, this would only require opening a particular port, when running the neurodesktop container, and setting up the sshd service in the neurodesktop image. Though I'm not sure if I can change these settings when starting up the NeurodeskApp.

**Reply by stebo85** [reply]
@m13slash9 - VS code just makes this very easy, but if you do not want to use VS code you can also connect to a running container using the following (you don't need SSH for that!):

```
docker ps
# note the name of the running container, e.g. neurodeskapp-49977

# now connect to this container
docker exec -ti neurodeskapp-49977 bash

# then switch to the jovyan user
su jovyan
```
Is this what you mean? Does this work for you? This would also work via SSH, where you run the neurodesktop container on a remote system.

**Reply by m13slash9** [reply]
Thank you! This is very close, and it seems to be the way I'd need to work with.

**Reply by stebo85** [reply]
cool - great to hear this works for you :)

**Comment by m13slash9** [comment]
Finally found a solution to enable proper SSH connection to neurodesktop.
If anyone interested, it is possible like this:

You first need to start neurodesktop manually with an additional port open for ssh (e.g., port 22222). 

To do this follow the instructions according to your system (e.g., [Windows](https://www.neurodesk.org/docs/getting-started/neurodesktop/windows/#2-run-neurodesktop)). 

Use the `Option 2 (Advanced): Using Terminal` set of commands, but add an additional port open in the `docker run` line, e.g.,

`docker run --shm-size=1gb -it --privileged --user=root --name neurodesktop -v C:/neurodesktop-storage:/neurodesktop-storage --mount source=neurodesk-home,target=/home/jovyan -p 8888:8888 -p 22222:22222 -e NEURODESKTOP_VERSION=2024-05-25 vnmd/neurodesktop:2024-05-25`

Then you need to set up the sshd service inside the neurodesk container. To do this, connect to the running container via
`docker exec -it -u root neurodesktop bash`

Then, open the `/etc/ssh/sshd_config` and change the default setting of `#Port 22` to your selected port e.g., `Port 22222`

Add your public authorization keys to `/home/jovyan/.ssh/authorized_keys`

Start the sshd service with `service ssh restart`

Now you should be able to SSH into neurodesktop with `ssh -p 22222 jovyan@localhost`

Note, that after stopping the container you'd be able to start it with just `docker start -ai neurodesktop`, but you will still need to restart the ssh service, e.g., with `docker exec -it -u root neurodesktop service ssh restart`

**Reply by stebo85** [reply]
Great to hear this is working :) have you tried using the default port 22? Does this work as well?

**Reply by m13slash9** [reply]
As far as I got it, starting Windows 10, there is a native sshd service in Windows that monitors port 22, then there is sshd service in WSL that monitors port 2222, so you'd need to either disable them and the associated routing, or just choose a different port.

Another possibility is to run docker with `-p 22222:22` port routing, so that the `ssh -p 22222 jovyan@localhost` would be directed to port 22 inside the container, but I have not checked that. It should remove the need to edit `/etc/ssh/sshd_config` but you'll still need to start the ssh service

**Reply by stebo85** [reply]
Cool - doing the port mapping during the docker call sounds like the best solution and I can't see why this wouldn't work :)

**Reply by m13slash9** [reply]
Just checked - it indeed works with `-p 22222:22` forwarding, so no need to edit `/etc/ssh/sshd_config`

**Reply by stebo85** [reply]
Great - thank you for testing and sharing the solution
